### ISSUE TITLE
Correlation between Batch Size & Performance in MP

### ISSUE THREAD

**Issue by iamyoungjo** [clab/dynet#503]
Hello @neubig @pmichel31415 ,

Currently I am spending time on tuning the model with different batch sizes (which I used as "num_iterations" argument in rnnlm-mp example) and shows the result as below graph:
![image](https://cloud.githubusercontent.com/assets/25552701/25586485/61103b56-2e54-11e7-9a96-9f0bb4d7b54f.png)

My implementation is done with Learner (based on the rnnlm-mp example).
Question is, why the batch-size difference causes difference in CPU based training like this?
I'm aware of that this question will be vague as I'm not providing the size of training or dev data but still I think it's possible to know more about what's causing the difference and correlation of these. If there's something that I need to understand on the implementation in mp.h, mp.cpp in order to narrow down this matter, please let me know.

Thanks,
YJ

**Comment by neubig** [comment]
I'm not sure what the question is... Batch size is a hyperparameter that affects accuracy, and it looks like your results are demonstrating this. This looks like normal behavior?

**Comment by iamyoungjo** [comment]
@neubig , Yes. Batch size is a hyper-parameter when you're training the model with GPU as it takes one batch and affects to parameters all at once. As far as I understood, however, in the implementation of mp.cpp and mp.h, it just separates the work by the number of workers (threads) and go by example by example. That's why I started having curiosity on this result as I set the number of thread fixed. As I read in the description of the multiprocess in dynet, I got the idea that it will average the backpropagation by the number of workers, but I would like to understand how exactly it calculates that.

**Comment by iamyoungjo** [comment]
@neubig , if there's anything I misunderstood, let me know.

**Comment by neubig** [comment]
I didn't write the code so I'm not 100% sure of the details, but basically it accumulates gradients until you've processed `samples_processed % batch_size == 0`. You can take a look at the code around here if you want more details: https://github.com/clab/dynet/blob/master/dynet/mp.h#L335
